src/System.IO/src/System/IO/BufferedStream.cs

```diff
@@ -0,0 +1,682 @@
+                    FlushWrite();
+
+                return _stream.Length;
+            }
+        }
+
+        public override Int64 Position
+        {
+            get
+            {
+                EnsureNotClosed();
+                EnsureCanSeek();
+
+                Contract.Assert(!(_writePos > 0 && _readPos != _readLen), "Read and Write buffers cannot both have data in them at the same time.");
+                return _stream.Position + (_readPos - _readLen + _writePos);
+            }
+            set
+            {
+                if (value < 0)
+                    throw new ArgumentOutOfRangeException("value", SR.ArgumentOutOfRange_NeedNonNegNum);
+                Contract.EndContractBlock();
+
+                EnsureNotClosed();
+                EnsureCanSeek();
+
+                if (_writePos > 0)
+                    FlushWrite();
+
+                _readPos = 0;
+                _readLen = 0;
+                _stream.Seek(value, SeekOrigin.Begin);
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                if (disposing && _stream != null)
+                {
+                    try
+                    {
+                        Flush();
+                    }
+                    finally
+                    {
+                        _stream.Dispose();
+                    }
+                }
+            }
+            finally
+            {
+                _stream = null;
+                _buffer = null;
+
+                // Call base.Dispose(bool) to cleanup async IO resources
+                base.Dispose(disposing);
+            }
+        }
+
+        public override void Flush()
+        {
+            EnsureNotClosed();
+
+            // Has WRITE data in the buffer:
+            if (_writePos > 0)
+            {
+                FlushWrite();
+                Contract.Assert(_writePos == 0 && _readPos == 0 && _readLen == 0);
+                return;
+            }
+
+            // Has READ data in the buffer:
+            if (_readPos < _readLen)
+            {
+                // If the underlying stream is not seekable AND we have something in the read buffer, then FlushRead would throw.
+                // We can either throw away the buffer resulting in data loss (!) or ignore the Flush.
+                // (We cannot throw becasue it would be a breaking change.) We opt into ignoring the Flush in that situation.
+                if (!_stream.CanSeek)
+                    return;
+
+                FlushRead();
+
+                // User streams may have opted to throw from Flush if CanWrite is false (although the abstract Stream does not do so).
+                // However, if we do not forward the Flush to the underlying stream, we may have problems when chaining several streams.
+                // Let us make a best effort attempt:
+                if (_stream.CanWrite || _stream is BufferedStream)
```

Comment:
Doesn't make a lot of sense, no.

For that matter, I don't follow this comment either. We note that the abstract stream doesn't throw on flush if CanWrite is false so we should therefore call Flush, but then we don't do that. Earlier it mentions we can't throw because it would make a breaking change, but why not try/catch the underlying stream's flush call?